tardis/spectrum/formal_integral/source_function.py

```diff
@@ -0,0 +1,285 @@
+
+        att_S_ul : np.ndarray
+            Attenuated source function
+        Jredlu : np.ndarray
+        Jbluelu : np.ndarray
+        e_dot_u : pd.DataFrame
+        """
+
+        # Parse states for required values
+        v_inner_boundary_index = sim_state.geometry.v_inner_boundary_index
+        v_outer_boundary_index = sim_state.geometry.v_outer_boundary_index
+        no_of_shells = sim_state.no_of_shells
+        dilution_factor = sim_state.dilution_factor
+        time_explosion = sim_state.time_explosion
+        volume = sim_state.volume
+
+        tau_sobolev = opacity_state.tau_sobolev
+        transition_probabilities = opacity_state.transition_probabilities
+
+        j_blue_estimator = transport_state.radfield_mc_estimators.j_blue_estimator
+        Edotlu_estimator = transport_state.radfield_mc_estimators.Edotlu_estimator
+        time_of_simulation = transport_state.packet_collection.time_of_simulation * u.s
+
+
+        # slice for the active shells
+        local_slice = slice(v_inner_boundary_index,v_outer_boundary_index)
+
+        transition_probabilities = transition_probabilities[:, local_slice]
+        tau_sobolevs = tau_sobolev[:, local_slice]
+
+        macro_ref = atomic_data.macro_atom_references
+        macro_data = atomic_data.macro_atom_data
+
+        no_lvls = len(levels)
+        no_shells = len(dilution_factor)
+
+        # Calculate e_dot_u
+        upper_level_index = atomic_data.lines.index.droplevel(
+            "level_number_lower"
+        )
+        e_dot_u = calculate_edotu(time_of_simulation, volume, tau_sobolevs, Edotlu_estimator, 
+                        macro_data, macro_ref,
+                        transition_probabilities, upper_level_index, no_of_shells, no_lvls,
+                        line_interaction_type=self.line_interaction_type
+                        )
+
+
+        # Calculate att_S_ul
+        transition_type = atomic_data.macro_atom_data.transition_type
+        transitions = atomic_data.macro_atom_data[transition_type == -1].copy()
+        transitions_index = transitions.set_index(
+            ["atomic_number", "ion_number", "source_level_number"]
+        ).index.copy()
+        transition_line_id = transitions.transition_line_id.values
+        lines = atomic_data.lines.set_index('line_id') 
+        lines_idx = lines.index.values 
+        
+        att_S_ul = calculate_att_S_ul(lines, transition_probabilities, 
+                            no_of_shells, transition_line_id, lines_idx, 
+                            transitions_index, transition_type,
+                            e_dot_u, time_explosion)
+
+        # Calculate Jredlu and Jbluelu
+        Jbluelu = calculate_Jbluelu(time_explosion, time_of_simulation, volume, j_blue_estimator)
+        Jredlu = calculate_Jredlu(Jbluelu, tau_sobolevs, att_S_ul)
+
+        return SourceFunctionState(att_S_ul, Jredlu, Jbluelu, e_dot_u)
+
+
+def calculate_edotu(time_of_simulation, volume, tau_sobolevs, Edotlu_estimator, 
```

Comment:
Could this be a method of the source function solver? I'm not sure when it would be needed elsewhere.